src/tribler/core/upgrade/tribler_db/decorator.py

```diff
@@ -0,0 +1,51 @@
+import functools
+import logging
+from typing import Callable
+
+from tribler.core.components.database.db.tribler_database import TriblerDatabase
+from tribler.core.utilities.pony_utils import db_session
+
+MIGRATION_METADATA = "_tribler_db_migration"
+
+logger = logging.getLogger('Migration (TriblerDB)')
+
+
+def migration(execute_only_if_version: int):
+    """ Decorator for migration functions.
+
+    The migration executes in the single transaction. If the migration fails, the transaction is rolled back.
+    The decorator also sets the metadata attribute to the decorated function. It could be checked by
+    calling the `has_migration_metadata` function.
+
+    After the successful migration, db version is set to the value equal to `execute_only_if_version + 1`.
+
+    Args:
+        execute_only_if_version: Execute the migration only if the current db version is equal to this value.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        @db_session
+        def wrapper(db: TriblerDatabase, **kwargs):
+            target_version = execute_only_if_version
+            if target_version != db.version:
+                logger.info(
+                    f"Function {func.__name__} is not executed because DB version is not equal to {target_version}. "
+                    f"The current db version is {db.version}"
+                )
+                return None
```

Comment:
It looks dangerous to silently skip the migration in case of a version mismatch and can lead to incorrect migrations.

As a possible example:
* We have migration chain `[a, b, c]`, and after the migration `c`, the database has version 3
* Two developers make two independent PRs with migrations `d` and `e`, both require db version 3
* After both PRs were successfully merged, the merge conflict in the migration chain was resolved to `[a, b, c, d, e]`, but both `d` and `e` still mistakenly require version 3 of the database.
* The execution of the migration chain silently skips the migration `e`

For that reason, it is better to raise an error instead of skipping the migration so the database will not be left in a surprising state.

Reply:
If we were to raise an exception as you suggested, then almost all runs would result in unsuccessful execution.

Example:

1. We have a migration chain [a, b, c] with the corresponding versions [1, 2, 3]. Let's represent them as [a(1), b(2), c(3)].
2. Our DB currently has a version set to 2.
3. We begin executing the migration chain starting with migration a(1).
4. This results in an exception because the DB version is not equal to 1.

Reply:
We can first select suitable migrations, as described in the previous comment, and execute only them. This way, the migration decorator can throw an exception if the migration was performed at an improper moment. It looks safer than silently skipping migrations with incorrect db version numbers.

Also, consider the migrations [a, b, c, d, e] that mistakenly expect the schema version [1, 3, 2, 4, 5]. With silent skipping of the incorrect migrations, only migrations 1 and 2 will be executed. It is better to prevent such errors.

Reply:
Actually, I'm not sure that it is a good idea to automatically fix **potential merge conflict errors**.

Also, I'm not sure that the Migration Chain should have a monopoly on changing the version value.

Reply:
> Actually, I'm not sure that it is a good idea to automatically fix potential merge conflict errors.

I agree, and what I like in the current approach is that the merge conflicts for migrations always require explicit conflict resolution, as migrations are explicitly listed in the same list.

What I'm concerned about is that the current PR approach allows silent skipping of migration if the schema number specified in the decorator does not agree with the order of migrations in the explicit list. Raising an error looks like a safer approach.